### PR TITLE
[BEAM-1287] Add support for PipelineOptions parameters

### PR DESCRIPTION
**Pull request by kennknowles**
This is a step towards eliminating catch-all context parameters and
making DoFns express their fine-grained data needs.

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

I'm letting this JIRA scope-creep to anything that helps clarify the parameters here. So this gives direct `PipelineOptions` access, and then we'll want direct `Element` access and `OutputReceiver` access and `WindowedOutputReceiver` access. If we jump right to the last one, we'll have redundant methods in the context and output receiver, which could be a smidge confusing.

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @bjchambers care to do a little review?

**Comment by kennknowles** [comment]
R: -@bjchambers
R: @tgroh I think the next closest person to this stuff?

**Comment by kennknowles** [comment]
Rebased and resolved conflicts. Please take a look.

**Comment by kennknowles** [comment]
R: @jkff I believe @tgroh is preoccupied - do you have time to look this over? I know you have been in this code to implement SDF.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignature.java [hunk 1]
```diff
@@ -306,6 +321,15 @@ public static TimerParameter timerParameter(TimerDeclaration decl) {
     }
 
     /**
+     * Descriptor for a {@link Parameter} of a subtype of {@link PipelineOptions}.
+     */
+    @AutoValue
+    public abstract static class PipelineOptionsParameter extends Parameter {
+      PipelineOptionsParameter() {}
+      public abstract TypeDescriptor<? extends PipelineOptions> pipelineOptionsT();
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignature.java]: Is this necessary - can we just provide a parameter of type exactly `PipelineOptions`? It's trivial for the user to convert it to the needed type with `.as()`, and in some cases the user may need to `.as()` it to several different classes, which makes the current implementation overkill.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/reflect/DoFnSignaturesTest.java [hunk 2]
```diff
@@ -329,6 +330,34 @@ public void onTimer(BoundedWindow w) {}
...
+  public void testPipelineOptionsSubclassParameter() throws Exception {
+    DoFnSignature sig =
+        DoFnSignatures.getSignature(new DoFn<String, String>() {
+          @ProcessElement
+          public void process(ProcessContext c, SubPipelineOptions options) {}
+        }.getClass());
+
+    assertThat(
+        sig.processElement().extraParameters(),
+        Matchers.<Parameter>hasItem(instanceOf(Parameter.PipelineOptionsParameter.class)));
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/reflect/DoFnSignaturesTest.java]: This should verify that pipelineOptionsT() is SubPipelineOptions.

**Comment by kennknowles** [comment]
I agree with the simplification. I'm not sure why I tried to get fancy.

**Pull request closed without merging**